lib/jormungandr/src/Cardano/Wallet/Jormungandr/Rewards.hs

```diff
@@ -0,0 +1,169 @@
+
+    , rEpochStart :: Word32
+        -- ^ indicates when this contribution start. note that if the epoch is
+        -- not the same or after the epoch_start, the overall contribution is zero.
+
+    , rEpochRate :: Word32
+        -- ^ the rate at which the contribution is tweaked related to epoch.
+
+    } deriving (Show, Eq, Generic)
+
+instance NFData RewardParams
+
+data Ratio = Ratio Word64 Word64
+    deriving (Show, Eq, Generic)
+
+ratio :: Ratio -> Double
+ratio (Ratio num den) = fromIntegral num / fromIntegral den
+
+instance NFData Ratio
+
+data RewardFormula
+    = HalvingFormula RewardParams
+    | LinearFormula RewardParams
+    deriving (Show, Eq, Generic)
+
+instance NFData RewardFormula
+
+data TaxParameters = TaxParameters
+    { taxFixed :: Word64
+        -- ^ A fix value taken from the total
+    , taxRatio :: Ratio
+        -- ^ An extra percentage taken from the total
+    , taxLimit :: Maybe Word64
+        -- ^ It is possible to add a max bound to the total value taken at each
+        -- cut.
+    } deriving (Generic, Eq, Show)
+
+instance NFData TaxParameters
+
+-- | limit the epoch total reward drawing limit to a portion of the total
+-- active stake of the system.
+--
+-- for example, if set to 10%, the reward drawn will be bounded by the
+-- 10% of the total active stake.
+--
+-- this value is optional, the default is no reward drawing limit
+data RewardLimit
+    = RewardLimitNone
+    | RewardLimitByAbsoluteStake Ratio
+    deriving (Show, Eq, Generic)
+
+instance NFData RewardLimit
+
+-- | settings to incentivize the numbers of stake pool to be registered
+-- on the blockchain.
+--
+-- These settings does not prevent more stake pool to be added. For example
+-- if there is already 1000 stake pools, someone can still register a new
+-- stake pool and affect the rewards of everyone else too.
+--
+-- if the threshold is reached, the pool doesn't really have incentive to
+-- create more blocks than 1 / set-value-of-pools % of stake.
+--
+-- this value is optional, the default is no pool participation capping
+data PoolCapping = PoolCapping
+    { minParticipation :: Word32
+    , maxParticipation :: Word32
+    } deriving (Show, Eq, Generic)
+
+instance NFData PoolCapping
+
+rewardsAt
+    :: (RewardLimit, Quantity "lovelace" Word64)
+    -> TaxParameters
+    -> EpochNo
+    -> RewardFormula
+    -> Quantity "lovelace" Word64
+rewardsAt limit tax epochNo = q . taxCut tax . capDrawing limit . \case
+    LinearFormula  params | ep < rEpochStart params -> 0
+    LinearFormula  params -> linearAbsorption params
+    HalvingFormula params | ep < rEpochStart params -> 0
+    HalvingFormula params -> halvingAbsorption params
+  where
+    q :: Double -> Quantity "lovelace" Word64
+    q = Quantity . round
+
+    ep = fromIntegral (unEpochNo epochNo)
+
+    linearAbsorption RewardParams{rFixed,rRatio,rEpochStart,rEpochRate} =
+        if a > c then 0 else c - a
+      where
+        a = r * (n / e)
```

Comment:
I think there should be a `floor` here.

See: https://github.com/input-output-hk/chain-libs/blob/149e01cf61a9eef2660a6944fe0acc73d490e70f/chain-impl-mockchain/doc/incentives.md#L105-L117

Reply:
_Looking more into other things first_

Reply:
You are pointing to the linear formula but the example from your link is about the halving formula ?

Reply:
Yet, these are nice examples to use as test vectors perhaps.